### ISSUE TITLE
What are the policies for rapidly changing modules and non-standard release versions?

### ISSUE THREAD

**Issue by aaronmondal** [bazelbuild/bazel-central-registry#205]
Most of our modules only change weekly or monthly, but sometimes we cut several releases in a day. Are there any guidelines on how frequently things can change in the central registry?

Also, what is the policy on using non-tag/non-release versions of repos? We use aggressively upstream compiler toolchains, so downstream repos tend to frequently raise deprecation warnings or simply fail to build. We currently work around these issues by vendoring our own modules cut from the patches that fix our issues.

We now want to try to be more in sync with the central registry, but the versions of some packages do not yet contain the fixes we need.

Do we need to maintain our non-tag/non-release modules separately from the central registry (i.e. in our own fork), or is it possible for us to upstream these modules here? For instance, `abseil-cpp` will raise a million warnings when built with clang 16 (probably 15 as well). The fixes for these warnings have already been merged in abseil main, but we do not want to wait until the next abseil release is cut.

Enabling more than standard releases would likely require versioning like `0.0.0-yyyymmdd.n-<commit-hash>` (where `n` is 0, 1, 2 etc if e.g. a hotfix is pushed in the same day) or similar. I've seen that some modules already are tagged like this (e.g. the `upb` module). What would be the compatibility level increment for releasing like that? I could imagine something like `compatibility_level=yyyymmdd`?

**Comment by Wyverald** [comment]
cc @meteorcloudy 

I don't believe we have a policy for frequent releases right now. Preferably only somewhat stable versions should be submitted to the central registry, but given the existence of "live at head" projects such as Abseil and upb, I think having unreleased/pre-release versions is unavoidable. The good thing is that we do support pre-release versions as denoted in the semver spec, so for `abseil-cpp` for example, you could create a "fake" version like "20221005-bcr.1" if necessary (let's try to keep the version format consistent within each module, as different projects have different versioning schemes).

> Most of our modules only change weekly or monthly, but sometimes we cut several releases in a day.

What are these modules that you cut several releases per day for? My gut take is that we should at least attempt to limit the number of versions in the BCR, say with some arbitrary frequency cap (just throwing "1 per day" out there).

**Comment by aaronmondal** [comment]
@Wyverald Dependencies for [`rules_ll`](https://github.com/eomii/rules_ll) and rules_ll itself. Usually we only release like once or twice a month, but we've had a case before where we had to hotfix issues like 3 times within a few hours :sweat_smile: This is mostly not the case though, and a theoretical max cap like once a day is completely fine for us.

With `rules_ll` specifically, I'm not sure whether that should be in the BCR anyways, since we use custom overlays for things like abseil and I don't think it would be wise to have duplicates of all the C++ modules in the BCR just because we prefer our own API over `rules_cc`.

The bigger issue is the LLVM project, which I'm tracking in #206. But even there once a day would be way more often than what we can realistically keep up with continuously. I just wanted to make sure that IF there is a case where we have to hotfix something in some exceptional case that we have "headroom" for that.

So as long as there is no such rule like "1 per month max and only stable versions" things are totally fine for us :blush:

> "fake" version like "20221005-bcr.1" if necessary (let's try to keep the version format consistent within each module, as different projects have different versioning schemes).

I really like that versioning method. I'll start getting some PRs ready :smile:

**Comment by Wyverald** [comment]
OK, so it sounds like the multiple-releases-per-day thing is expected to be the exception, not the norm, in which case I'm totally okay with this.

Also note that we don't intend every single released version of a project to be available in BCR, so you could also wait for a day before submitting a new release into BCR ("baking" it, if you will), if you really need to avoid the frequent updates.

**Comment by meteorcloudy** [comment]
Just want to mention that  PRs still take some time to review (possibly longer than 1 day), so even if you have multiple hot fixes for a release, it's likely that when we receive the PR for the last release we can just discard previous unmerged PRs.

I do know a use case that the project needs to depend on a specific commit and update the dependency multiple times in a day.  TensorFlow: https://github.com/tensorflow/tensorflow/commits/master/third_party/llvm/workspace.bzl

In that case, it's probably easier for them to just use overrides (or just introduce the same dependency with a module extension) so that they won't get blocked by the BCR process.

**Comment by aaronmondal** [comment]
Closing as my questions have been answered :relaxed:
